refactoring of QgsMapToolCapture to integrate shape map tools

3nids opened the pull request:
This moves part of the code from `QgsMapToolDigitizeFeature` to `QgsMapToolCapture` so the tool can actually capture point, line and polygons. It's mainly the 'cadCanvasReleaseEvent` which has been transfered.

This is a preliminary work to support shape digitizing in more map tools (like add part, add ring, fill ring or anything else).

The idea is that `QgsMapToolCapture` should actually perform the digitization of point/line/polygon and let subclasses deal to what they want to do with it (create feature, add part, add ring, etc).

Then, `QgsMapToolCapture` will be able to digitize shapes (by adding a new enum entry `CaptureTechnique::ShapeDigitizing`). The idea is that the shape toolbar will be turned into a checkable choice so that `QgsMapToolCapture` will be able to perform any of the capture technique entries:
  * StraightSegments
  * CircularString
  * Streaming
  * Shapes
 
I'd like to get opinions of the map tools experts @lbartoletti @nyalldawson. If you consider I should open a QEP for this discussion, let me know.

3nids committed:
make QgsMapToolCapture capable of capturing point/line/polygons

This moves part of the code from QgsMapToolDigitizeFeature to QgsMapToolCapture so the tool can actually capture point, line and polygons. It's mainly the 'cadCanvasReleaseEvent` which has been transfered.

nyalldawson commented:
I haven't looked at the code, but I fully support the design changes here. (I've wanted similar to allow the shape digitising for annotation items!). Great work @3nids!

3nids committed:
use a current CaptureTechnique in QgsMapToolCapture

QgisApp has been adapted to switch between the different techniques

lbartoletti commented:
I can't review it this week, but it's a big +1 for this design. Thanks @3nids !

3nids committed:
add point/line/polygon specific handlers for capture map tool

3nids committed:
convert add part map tool to use QgsMapToolCapture capabilities

3nids commented:
thanks @lbartoletti and @nyalldawson for the thumbs up

I have gone a bit furthe:
* `QgsMapToolCapture` now has a capture technique defined
* ground work to introduce the shape technique is done
* add part has been turned to fully use QgsMapToolCapture (by using the parent cadCanvasReleaseEvent)

3nids committed:
fix use of deprecated methods

3nids committed:
also create a virtual handler for QgsMapToolDigitizeFeature::featureDigitized

nyalldawson commented on src/gui/qgsmaptoolcapture.cpp:
```diff
@@ -1073,3 +1108,191 @@ void QgsMapToolCapture::updateExtraSnapLayer()
+    catch ( QgsCsException &cse )
+    {
+      Q_UNUSED( cse )
+      emit messageEmitted( tr( "Cannot transform the point to the layers coordinate system" ), Qgis::MessageLevel::Warning );
```
```suggestion
      emit messageEmitted( tr( "Cannot transform the point to the layer's coordinate system" ), Qgis::MessageLevel::Warning );
```

nyalldawson commented on src/gui/qgsmaptoolcapture.cpp:
```diff
@@ -1073,3 +1108,191 @@ void QgsMapToolCapture::updateExtraSnapLayer()
 
+
+
+void QgsMapToolCapture::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
```
Why is this all moved to the base class? There's a lot here which isn't generic logic, e.g. handling the avoid intersections behaviour, force segmentisation if the layer doesn't support curves, etc. This logic will break existing subclasses, e.g. the annotation item digitization tools which don't have a vector layer associated with them.

3nids replied:
My idea is also that things like intersections could be optional on demand.

But having everything in the base class will benefit to other map tools (add part, add/fill ring).

nyalldawson replied:
Hmm... I think this should be at least opt in only. But really, I don't think it has any place in a base QgsMapToolCapture class, which can be used to capture ANY geometry for any purpose, where all the avoid intersection/etc logic is unwanted. E.g. imagine a (very theoretical) "measure area of shape" tool which allows you to draw a rect/circle/... and shows the corresponding area in the status bar... there's no way you'd want geometries drawn using this tool to magically trigger the avoid intersections logic!

If you want to keep this stuff reusable, I'd insert a new class between QgsMapToolCapture and QgsMapToolDigitizeFeature, allowing the QgsMapToolCapture class to be kept for pure geometry captures only without any changes to the project/layer states.

3nids replied:
I have added an intermediate class `QgsMapToolCaptureLayerGeometry` in 47c98e0636f0eef3d22063ecc23396c84f72bccd

3nids commented on src/gui/qgsmaptoolcapture.h:
```diff
@@ -320,6 +334,31 @@ class GUI_EXPORT QgsMapToolCapture : public QgsMapToolAdvancedDigitizing
+     * Called when the geometry is captured
+     * \since QGIS 3.24
+     */
+    virtual void geometryCaptured( const QgsGeometry &geometry ) {Q_UNUSED( geometry )} SIP_FORCE
```
I thought it was ok with forcing them with SIP_FORCE. Am I wrong?

nyalldawson replied:
I'm not sure either way -- it just looks problematic to me when a protected method is guaranteed to work correctly

3nids replied:
quoting https://stackoverflow.com/a/2170696/1548052

> This lets the derived classes override the function to customize the behavior as needed, without further exposing the virtual functions directly by making them callable by derived classes (as would be possible if the functions were just protected). The point is that virtual functions exist to allow customization; unless they also need to be invoked directly from within derived classes' code, there's no need to ever make them anything

And if you consider the idea of the additional class to handle the avoidIntersections, we'll be dealing with the private handler on the geometry and it will be better to just not forward the specific handlers.

3nids committed:
more dox

3nids commented on src/gui/qgsmaptoolcapture.h:
```diff
@@ -320,6 +334,31 @@ class GUI_EXPORT QgsMapToolCapture : public QgsMapToolAdvancedDigitizing
   private:
+
+    /**
+     * Called when the geometry is captured
```
I added more docs both to the class main doc and to the handler

3nids committed:
use const abstract geom in virtual handlers

3nids committed:
add new class QgsMapToolCaptureLayerGeometry to handle layer specific operation in capture map tool

such as avoiding intersections

3nids committed:
allow to add linear geometries on curved geometry layers

3nids committed:
make actions exclusive

3nids committed:
add settings registry to app

3nids committed:
add a registry for shape map tools

3nids committed:
abstract class for shape map tools

3nids committed:
adapt QgsMapToolCapture to correctly support shape map tools

3nids committed:
clean up of QgisApp

3nids committed:
new class QgsMapToolsDigitizingTechniqueManager to handle actions in app related to capture map tools

3nids committed:
clean up QgisInterface

3nids committed:
refactoring of existing shape tools

3nids commented:
Some progress have been made:
* a registry of shape tools has been added 8649b64
* a new class `QgsMapToolsDigitizingTechniqueManager` handles actions in app related to capture map tools 3874352 (this allows cleaning up quite some code in QgisApp, see https://github.com/qgis/QGIS/pull/46687/commits/f5d9e90c1896b588319508b4bb62e92efe43c222#diff-a6c6864f28b601d260d5aff00f149039a0df622797f72f15311d052b1a38b77f )
* the refactoring of existing shape tools is on-going ( 0ec45e8 )

Remains:
* complete refactoring of shape tools
* complete refactoring of capture map tools (add feature and add part are done for now, remains add/fill ring, etc.)
* adapt existing shape tool tests to work with new code

I know this a big code change, but I would like if possible to merge it during this cycle.

3nids committed:
sipify

3nids commented:
Refactoring of existing shape tools is completed

3nids committed:
refactor add ring to fully support capture map tool

3nids commented:
And here are the curved rings:
![image](https://user-images.githubusercontent.com/127259/149410442-3a17eecd-0dd4-4987-98b5-ccfec7ea07f1.png)

3nids committed:
add missing folder to Doxygen

3nids committed:
fix layout

3nids committed:
fix erasing at iterator pos

3nids committed:
fix unused warning

3nids committed:
fix more dox

3nids committed:
fix cpp check warning

3nids committed:
fix unused warning

3nids commented:
![ezgif-7-41239961b7](https://user-images.githubusercontent.com/127259/149415291-3ebdabea-4726-4dad-baf1-c863b65eeb25.gif)

3nids committed:
fix annotation map tool does not support shape + set tool name

3nids committed:
correctly handle case when the capture is not done on a vector layer (annotation, mesh, …)

3nids committed:
enable shapes in annotation map tool

3nids committed:
correctly undo and clean

3nids committed:
adapt existing shape tests

the deletion test with circular vertices has been dropped since the capture map tool behaves differently

3nids committed:
fix warning

3nids committed:
refactor fill ring to support shape digitizing

3nids commented:
Tests are passing and this is completed.
This is ready for merge as far as I am concerned.

There seems to be an issue with the MingW build but I cannot access the log. It's maybe the QgsSettingsEntryEnumFlag<QgsMapToolCapture::CaptureTechnique>.

3nids committed:
fix win build

3nids commented:
I cannot find a way to have `QgsSettingsEntryEnum` to work on windows build. I propose to use an integer setting in the mean time to merge this and fix it later on.

3nids committed:
fix more tests

3nids committed:
avoid detach warnings

3nids committed:
fix app test + clean up

3nids committed:
harmonize new settings with existing ones

m-kuhn commented:
Shall we merge this soon to get the best of the testing phase still?

lbartoletti commented:
> Shall we merge this soon to get the best of the testing phase still?

I hope I'll have time next week to review the PR and test it.

3nids commented:
@lbartoletti @nyalldawson gentle ping 
I am a bit concerned by the feature freeze period flying by and that time would be short for testing.
Please let me know if a finer description of the code changes would help for a potential review. We can also arrange a call for this.

lbartoletti commented:
@3nids I plan to watch your PR in the afternoons this week.
I start tomorrow. However, I see some problem with the UI:

![image](https://user-images.githubusercontent.com/7521540/150982478-d77aa934-32b1-4775-9434-52094b17c624.png)

And I tried to add a circle part into a rectangle multilinestring, and I get an error.

I'll continue tomorrow.

lbartoletti commented on python/gui/auto_generated/qgisinterface.sip.in:
```diff
@@ -672,69 +672,133 @@ Returns the Hide Deselected Layers action.
 
-    virtual QAction *actionCircle2Points()  = 0;
+
+ virtual QAction *actionCircle2Points() /Deprecated/;
```
Maybe add a documentation to explain how to launch the maptools via iface (same for the others).

lbartoletti replied:
These actions are used in some plugins (was asked for qgep/qwat, and after used by other tools)

3nids replied:
ok for you like this bf44303 ?

lbartoletti replied:
Yes, thanks!

3nids committed:
fix categories

3nids committed:
support adding multi lines as a part

3nids committed:
fix adding curve part to multi line

3nids commented:
> @3nids I plan to watch your PR in the afternoons this week. I start tomorrow. However, I see some problem with the UI:
> 
> ![image](https://user-images.githubusercontent.com/7521540/150982478-d77aa934-32b1-4775-9434-52094b17c624.png)

fixed in e8575bdd04808a7b7a6f971fd102d5a50f052801

> 
> And I tried to add a circle part into a rectangle multilinestring, and I get an error.
> 
> I'll continue tomorrow.

QgsGeometryEditUtils::addPart was not supporting adding curve part to multi-line string.
fixed in 7d32a458b4e456c27d208975a4c4e1ed67530870

3nids committed:
also handle points

3nids committed:
code a bit clearer

3nids committed:
cast not always valid

3nids committed:
allow adding curved polygon to multipolygon

3nids committed:
add test for QgsGeometry::addPart with curved parts on non-curved geoms (lines and polygons)

3nids commented:
The same issue you spotted but for polygons (adding curved part to non-curved geoms) has been fixed too.
For the 2 cases, tests have been added bab578b6902817144a974ae5e06d810a16b3e6fb

3nids committed:
fix with Python < 3.9

3nids committed:
better dox for deprecated interface actions methods

3nids committed:
remove files leftover

3nids committed:
remove leftover circular string curve point tool

3nids commented:
Considering the freeze window getting shorter, I would like to merge this PR by the end of weekend.
If there are remarks post-merge, I would obviously be involved to take them into account and potentially adapt the code.

Is there any objection?

lbartoletti commented:
> Considering the freeze window getting shorter, I would like to merge this PR by the end of weekend. If there are remarks post-merge, I would obviously be involved to take them into account and potentially adapt the code.
> 
> Is there any objection?

Well, it is a considerable PR. I started to review it, but my spare time is very limited now. I hope to finish reviewing and test it this week, but no idea, if I can do it.

I prefer to wait and be sure of the behaviors rather than rush for 3.24.0.

But maybe other contributors have time to test your PR ( @uclaros @SrNetoChan @DelazJ etc.) ?

nyalldawson commented:
At this stage I'm -1. I'd rather delay it to 3.26 given the large scope of work and high chance of regressions.

3nids committed:
add default Z/M values when calling QgsGeometry::coerceToType

3nids commented:
just rebased with integrated changes from #47039

3nids commented:
I understand the concerns.

I'd like to bring some precisions about the relatively limited scope of the code changes (compared to the number of line changes):
* the core changes are quite limited
  * a new registry of shape map tools (not yet exposed to Python) 713243108ac857437f35f4697c432064934112e0
  * a base class for shape map tools + metadata (not yet exposed to Python) 55e8819240a19e1bf94abe66e2c3855bebbced5a
  * `QgsMapTool:: toMapCoordinates` has bee made public [0aed1cc](https://github.com/qgis/QGIS/commit/0aed1cc6604ac0e7e29d4ccaad402d2b2054cb73#diff-d7424aca404fdf8b427c4900aef6b8457fde02394c8dd2ead4f35392e6a67519R265)
  * `QgsMapToolEdit:: createGeometryRubberBand` has been made public [0aed1cc](https://github.com/qgis/QGIS/commit/0aed1cc6604ac0e7e29d4ccaad402d2b2054cb73#diff-8bbefcf10c8512925beb1e20e554275772a08b85625a35e4f3f1dfcc8707b257R62)
* the canvas mouse event code from add feature tool was move to the parent class QgsMapToolCapture: 864787e63f0386e48b2cc6b57f5afb676cbd322a

* other digitizing tools (add/fill ring, add part) now take advantage of this refactoring
  * add part: https://github.com/qgis/QGIS/pull/46687/commits/e2e8a1285022478a257c61e92f130b7ad715a591
  * add ring: e2e8a1285022478a257c61e92f130b7ad715a591
  * fill ring: f476685310b94c0c0f1415a903484243b5ce9b4a
* the rest is mainly code refactoring of the former shape map tools 42d86f74e75d71491164813b9327216fb7e097c3

This PR makes digitizing techniques (curves, streaming and shape) now usable in more map tools. This is a major improvement towards consistency in the app.

It also brings fixes to the core method `QgsGeometryEditUtils::addPart` (a63af99a55b8d5dcc557be468390d1c906950b31 11e63e9db2a3769a714eed2b608422539031eda1 e83c8a5bb5bf85d2ea550df15faebb53926f2d78), for which tests were added (after a merged refactoring of the test #47009).

I originally thought that 2 weeks were left before release, while there are still 3. I am fully available until then for any potential follow-up.
I believe the benefit risk balance is still favorable for a merge at this time.

nyalldawson commented on src/gui/maptools/qgsmaptoolshaperegistry.cpp:
```diff
@@ -0,0 +1,76 @@
+  {
+    if ( ( *it )->id() == id )
+    {
+      it = mMapTools.erase( it );
```
should this also delete the tool? The registry has ownership, so I'd think so.

nyalldawson commented on src/gui/maptools/qgsmaptoolshaperegistry.h:
```diff
@@ -0,0 +1,104 @@
+    void addMapTool( QgsMapToolShapeMetadata *mapTool SIP_TRANSFER );
+
+    /**
+     * Removes a registered relation widget with given \a widgetType
```
make a note here on whether the tool is deleted

nyalldawson commented on src/gui/maptools/qgsmaptoolshaperegistry.h:
```diff
@@ -0,0 +1,104 @@
+    //! Returns the list of map tools
+    QList<QgsMapToolShapeMetadata *> mapToolMetadatas() const {return mMapTools;}
+
+    //! Returns the map tool metadata for the gin \a id
```
```suggestion
    //! Returns the map tool metadata for the given \a id
```

nyalldawson commented on src/gui/maptools/qgsmaptoolshaperegistry.h:
```diff
@@ -0,0 +1,104 @@
+    QgsMapToolShapeMetadata *mapToolMetadata( const QString &id ) const;
+
+    /**
+     * Constructs the map tool at the given \a id for the given \a parentTool
```
```suggestion
     * Constructs the map tool at the given \a id for the given \a parentTool
     * Caller takes ownership of the returned tool.
```

(although given that this isn't exposed to sip, we could potentially return a unique_ptr here. Or is exposing this to sip in your long term vision?)

3nids replied:
yes, that is the plan. Just waiting a bit to be sure about the API.

nyalldawson commented on src/gui/maptools/qgsmaptoolshaperegistry.h:
```diff
@@ -0,0 +1,104 @@
+    virtual QgsMapToolShapeAbstract::ShapeCategory category() const = 0;
+
+    //! Creates the shape map tool for the given \a parentTool
+    virtual QgsMapToolShapeAbstract *factory( QgsMapToolCapture *parentlTool ) const = 0;
```
```suggestion
    virtual QgsMapToolShapeAbstract *factory( QgsMapToolCapture *parentlTool ) const SIP_FACTORY = 0;
```

nyalldawson replied:
Also, "create" would be a more standard-QGIS API terminology for this action.

nyalldawson commented on src/gui/maptools/qgsmaptoolshaperegistry.h:
```diff
@@ -0,0 +1,104 @@
+    //! Returns the shape category of the tool
+    virtual QgsMapToolShapeAbstract::ShapeCategory category() const = 0;
+
+    //! Creates the shape map tool for the given \a parentTool
```
Make a note about ownership here

nyalldawson commented on src/gui/maptools/qgsmaptoolshapeabstract.h:
```diff
@@ -0,0 +1,117 @@
+
+/**
+ * \ingroup gui
+ * \brief QgsMapToolShapeAbstract is a base class for shape map tools to be use in QgsMapToolCapture
```
```suggestion
 * \brief QgsMapToolShapeAbstract is a base class for shape map tools to be used by QgsMapToolCapture.
```

nyalldawson commented on src/gui/maptools/qgsmaptoolshapeabstract.h:
```diff
@@ -0,0 +1,117 @@
+      Circle,//!< Circle
+      Ellipse,//!< Ellipse
+      Rectangle,//!< Rectangle
+      RegularyPolygon,//!< RegularyPolygon
```
```suggestion
      RegularPolygon,//!< RegularPolygon (e.g pentagons or hexagons)
```

nyalldawson commented on src/gui/maptools/qgsmaptoolshapeabstract.h:
```diff
@@ -0,0 +1,117 @@
+
+    /**
+     * Called for a mouse release event
+     * Must return true if the digitization has ended and the geometry is correctly set
```
```suggestion
     * Must return TRUE if the digitization has ended and the geometry is correctly set
```

nyalldawson commented on src/gui/maptools/qgsmaptoolshapeabstract.h:
```diff
@@ -0,0 +1,117 @@
+    virtual void cadCanvasMoveEvent( QgsMapMouseEvent *e, QgsMapToolCapture::CaptureMode mode ) = 0;
+
+    /**
+     * Eventually filters a key press event
```
eventually? I think there's a translation issue here

nyalldawson commented on src/gui/maptools/qgsmaptoolshapeabstract.cpp:
```diff
@@ -0,0 +1,49 @@
+  e->ignore();
+}
+
+void QgsMapToolShapeAbstract::clean()
```
should clean() be called in the destructor too?

nyalldawson commented on src/gui/maptools/qgsmaptoolcapturelayergeometry.h:
```diff
@@ -0,0 +1,68 @@
+/**
+ * \ingroup gui
+ * \brief QgsMapToolCaptureLayerGeometry is a base class for map tools digitizing layer geometries
+ * It will implement avoid intersections
```
```suggestion
 * This map tool subclass automatically handles intersection avoidance with other layers in the active project whenever a geometry is digitized by the user.
```

nyalldawson commented on src/gui/qgsmaptoolcapture.cpp:
```diff
@@ -1073,3 +1206,187 @@ void QgsMapToolCapture::updateExtraSnapLayer()
+      //does compoundcurve contain circular strings?
+      //does provider support circular strings?
+      const bool hasCurvedSegments = captureCurve()->hasCurvedSegments();
+      const bool providerSupportsCurvedSegments = vlayer && ( vlayer->dataProvider()->capabilities() & QgsVectorDataProvider::CircularGeometries );
```
This logic shouldn't be in the base class, but rather in the QgsMapToolCaptureLayerGeometry class.

nyalldawson commented on src/gui/annotations/qgscreateannotationitemmaptool_impl.cpp:
```diff
@@ -63,9 +63,17 @@ QgsMapToolCapture::Capabilities QgsMapToolCaptureAnnotationItem::capabilities()
 
-bool QgsMapToolCaptureAnnotationItem::supportsTechnique( CaptureTechnique ) const
+bool QgsMapToolCaptureAnnotationItem::supportsTechnique( CaptureTechnique technique ) const
 {
-  return true;
+  switch ( technique )
+  {
+    case CaptureTechnique::StraightSegments:
+    case CaptureTechnique::CircularString:
+    case CaptureTechnique::Streaming:
+      return true;
+    case CaptureTechnique::Shape:
+      return true;
+  }
```
(or just simplified!)

nyalldawson commented:
I'm happy with the general direction of this PR now @3nids -- a few changes to make based on the above review, but it's relatively minor stuff.

3nids committed:
Apply suggestions from code review

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

3nids committed:
fixes from review

3nids committed:
move layer specific part to specific tool

3nids committed:
fix typo

3nids committed:
fix leak

3nids commented:
Thanks a loooooot for the review @nyalldawson.
Comments have been addressed.

3nids committed:
fix dox

lbartoletti commented on python/gui/auto_generated/qgisinterface.sip.in:
```diff
@@ -672,69 +672,133 @@ Returns the Hide Deselected Layers action.
 %Docstring
 Returns the native add circle from 2 points action. Call :py:func:`~QgisInterface.trigger` on it to set the map tool.
+
+.. deprecated:: QGIS 3.24
```
```suggestion
.. deprecated:: QGIS 3.26
```

lbartoletti reviewed (approved):
Except for a few details, it's good for me.

3nids committed:
fix segmentization

3nids committed:
call map tool implementation of addCurve when adding trace curve to avoid point duplication

3nids committed:
call sub-class implementation

3nids committed:
fix since 3.24 -> 3.26

3nids committed:
fix test

3nids committed:
add test to avoid extra curves when using tracing

3nids commented on src/app/maptools/qgsmaptoolshapecircleabstract.cpp:
```diff
@@ -13,42 +13,23 @@
+void QgsMapToolShapeCircleAbstract::addCircleToParentTool()
 {
   if ( !mParentTool || mCircle.isEmpty() )
-  {
```
I personally like the readability when there is only one line.
The coding guidelines are not super explicit about this, but there is an example with it https://docs.qgis.org/3.16/en/docs/developers_guide/codingstandards.html#put-commands-on-separate-lines

3nids committed:
fix headers

agiudiceandrea commented:
Hi @3nids, could you please have a look at the issue https://github.com/qgis/QGIS/issues/49541 which it seems to me it is related to this PR.
